idaes/generic_models/costing/SSLW.py

```diff
@@ -0,0 +1,1493 @@
+        using Chemical Engineering base cost index of 500.
+
+        Purchase Cost = pressure_factor *
+                        material_factor *
+                        length_factor *
+                        base_cost_per_unit *
+                        number_of_units
+
+        Args:
+            hx_type - HXType Enum indicating type of heat exchanger design,
+                      default = HXType.Utube.
+            material_type - HXMaterial Enum indicating material of
+                            construction, default = HXMaterial.SS_SS.
+            tube_length - HXTubeLength Enum indicating length of HX tubes,
+                          default = HXTubeLength.TwelveFoot.
+            integer - whether the number of units should be constrained to be
+                      an integer or not (default = True).
+        """
+        # Validate arguments
+        if hx_type not in HXType:
+            raise ConfigurationError(
+                f"{blk.unit_model.name} received invalid argument for hx_type:"
+                f" {hx_type}. Argument must be a member of the HXType Enum.")
+        if material_type not in HXMaterial:
+            raise ConfigurationError(
+                f"{blk.unit_model.name} received invalid argument for "
+                f"material_type: {material_type}. Argument must be a member "
+                "of the HXMaterial Enum.")
+        if tube_length not in HXTubeLength:
+            raise ConfigurationError(
+                f"{blk.unit_model.name} received invalid argument for "
+                f"tube_length: {tube_length}. Argument must be a member "
+                "of the HXTubeLength Enum.")
+
+        # Build generic costing variables
+        _make_common_vars(blk, integer)
+
+        # Length correction factor
+        c_fl = {'8ft': 1.25, '12ft': 1.12, '16ft': 1.05, '20ft': 1.00}
+        blk.length_factor = pyo.Param(
+            mutable=True,
+            initialize=c_fl[tube_length],
+            doc='HX tube length correction factor')
+
+        blk.hx_oversize = pyo.Param(mutable=True,
+                                    initialize=1.1,
+                                    doc='HX oversize factor (1.1 to 1.5)',
+                                    units=pyo.units.ft**-2)
+
+        # --------------------------------------------------
+        # Base cost calculation - based on selected heat exchanger type:
+        alpha_dict = {
+            HXType.floating_head: {1: 11.9052, 2: 0.8709, 3: 0.09005},
+            HXType.fixed_head: {1: 11.2927, 2: 0.8228, 3: 0.09861},
+            HXType.Utube: {1: 11.3852, 2: 0.9186, 3: 0.09790},
+            HXType.kettle_vap: {1: 12.2052, 2: 0.8709, 3: 0.09005}}
+        alpha = alpha_dict[hx_type]
+
+        # Convert area to square feet
+        area_unit = (pyo.units.convert(blk.unit_model.area,
+                                       to_units=pyo.units.ft**2) /
+                     blk.number_of_units)
+
+        def hx_cost_rule(blk):
```

Comment:
Why use decorators for the Expression, but not here for the Constraint?

Reply:
That is how the original costing code was written, and I was trying to keep my effort to a minimum.